### PR TITLE
Aligning Point Cloud and Camera Frustums

### PR DESCRIPTION
**Pull request by adisingh50**
It's been a while since my last commit, but I realized my last PR was left in an unfinished state. Thus, I decided to make an entirely new PR which is up to date with all the new codebases changes made over the summer. Once this is merged, I'll remove the old, lingering PRs.

In this PR, I have modified `io.py` and added a new file `ellipsoid.py` which contains all the helper functions necessary to align the point cloud and camera frustums to the x,y,z axes using SVD. I've also added `test_ellipsoid_utils.py` to perform unit testing on the helper functions in `ellipsoid.py`.

After Point Cloud Alignment:
image

### CODE REVIEW

File: gtsfm/utils/io.py [hunk 1]
```diff
@@ -297,8 +303,10 @@ def write_images(gtsfm_data: GtsfmData, images: List[Image], save_dir: str) -> N
             iTw = camera.pose().inverse()
             iRw_quaternion = iTw.rotation().quaternion()
             itw = iTw.translation()
-            tx, ty, tz = itw
-            qw, qx, qy, qz = iRw_quaternion
+
+            iuprightRi, iupright_t_i = ellipsoid_utils.transform_camera_frustums(itw, iRw_quaternion, mean, wuprightRw)
```
Comment (johnwlambert) [inline gtsfm/utils/io.py:307]: thanks for the PR, @adisingh50.

Let's pull the orthogonal alignment out of the file I/O, and put it into the scene optimizer itself. So that the COLMAP writing function here is purely I/O, no computation.

Could you do that? Could you pull in master to this branch, also?

File: gtsfm/utils/ellipsoid.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    if point_cloud.shape[1] != 3:
+        raise TypeError("Point Cloud should be 3 dimensional")
+
+    mags = np.linalg.norm(point_cloud, axis=1)
+    cutoff_mag = np.percentile(mags, 95)
+    points_filtered = point_cloud[mags < cutoff_mag]
+    return points_filtered
+
+
+def get_rotation_matrix(point_cloud: np.ndarray) -> np.ndarray:
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:88]: i think the fn name is a bit vague -- could be more specific to ellipsoid fitting?

File: gtsfm/utils/ellipsoid.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+"""Algorithms to center and align 3D points and camera frusums to the x, y, and z axes using SVD. Used in React
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:1]: small typo on "frustums"

File: gtsfm/utils/ellipsoid.py (hunk 2 above)
Comment (adisingh50) [reply gtsfm/utils/ellipsoid.py:88]: Ok, I'll rename this to `get_alignment_rotation_matrix_from_svd`

File: gtsfm/utils/ellipsoid.py (hunk 3 above)
Comment (adisingh50) [reply gtsfm/utils/ellipsoid.py:1]: Oops. thanks

File: rtf_vis_tool/src/Components/frustum_classes/view_frustum.js [hunk 4]
```diff
@@ -127,11 +127,11 @@ class ViewFrustum {
...
-        var v0_worldframe = wTc.transform_from(v0_camframe);
-        var v1_worldframe = wTc.transform_from(v1_camframe);
-        var v2_worldframe = wTc.transform_from(v2_camframe);
-        var v3_worldframe = wTc.transform_from(v3_camframe);
-        var v4_worldframe = wTc.transform_from(v4_camframe);
+        var v0_worldframe = cTw.transform_from(v0_camframe);
+        var v1_worldframe = cTw.transform_from(v1_camframe);
+        var v2_worldframe = cTw.transform_from(v2_camframe);
+        var v3_worldframe = cTw.transform_from(v3_camframe);
+        var v4_worldframe = cTw.transform_from(v4_camframe);
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/view_frustum.js:134]: it was right before :- )

File: rtf_vis_tool/src/Components/frustum_classes/view_frustum.js [hunk 5]
```diff
@@ -111,11 +111,11 @@ class ViewFrustum {
         return scaled_ray_dirs
     }
 
-    get_mesh_vertices_worldframe(wTc) {
+    get_mesh_vertices_worldframe(cTw) {
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/view_frustum.js:114]: we should keep wTc as the argument, as it was before

File: rtf_vis_tool/src/Components/frustum_classes/se3.js [hunk 6]
```diff
@@ -20,16 +20,16 @@ class SE3 {
         if (translation.shape[0] !== 3) throw new Error('Invalid Translation Matrix');
         this.rotation = rotation;
         this.translation = translation;
-        this.transform_matrix = nj.identity(4);
+        this.cTw = nj.identity(4);
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/se3.js:23]: we can keep this more generic to represent any class, so I would prefer to call it what it was before

File: gtsfm/scene_optimizer.py [hunk 7]
```diff
@@ -230,6 +237,51 @@ def _filter_dict_keys(dict: Dict[Any, Any], ref_dict: Dict[Any,Any]) -> Dict[Any
         # return the entry with just the sfm result
         return output_graph[0]
 
+def align_gtsfm_data(gtsfm_data: GtsfmData) -> Delayed:
```
Comment (johnwlambert) [inline gtsfm/scene_optimizer.py:240]: so we'll need to apply this to the GT, and to the estimated pose graphs, jointly. I propose that we estimate the transformation using the ground truth, and then add a classmethod to GtsfmData that aligns given a Pose3 object. And then call that classmethod on the estimated GtsfmData and the GT GtsfmData

File: gtsfm/scene_optimizer.py [hunk 8]
```diff
@@ -216,6 +219,10 @@ def _filter_dict_keys(dict: Dict[Any, Any], ref_dict: Dict[Any,Any]) -> Dict[Any
         # Save metrics to JSON and generate HTML report.
         auxiliary_graph_list.extend(save_metrics_reports(metrics_graph_list))
 
+        # Modify BA input and BA output to have point clouds and frustums aligned with x,y,z axes.
+        ba_input_graph = dask.delayed(align_gtsfm_data)(ba_input_graph)
```
Comment (johnwlambert) [inline gtsfm/scene_optimizer.py:223]: let's estimate the transform just once, since we need to make sure these two pose graphs are perfectly aligned with each other

File: rtf_vis_tool/src/Components/frustum_classes/view_frustum.js (hunk 4 above)
Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/view_frustum.js:134]: Ok. I'll change it back to `wTc` like before

File: rtf_vis_tool/src/Components/frustum_classes/se3.js (hunk 6 above)
Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:23]: Cool. I'll change it back

File: gtsfm/scene_optimizer.py (hunk 7 above)
Comment (adisingh50) [reply gtsfm/scene_optimizer.py]: I've made the changes to estimate the alignment transformation based on `ba_output`, and then apply that to `ba_input` as well using a classmethod in GtsfmData. I will apply this transformation to the GT camera poses in the next PR, since this one is becoming faily lengthy already.

File: gtsfm/common/gtsfm_data.py [hunk 9]
```diff
@@ -387,3 +387,53 @@ def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "Gtsfm
             aligned_data.add_track(track_aligned)
 
         return aligned_data
+
+    @classmethod
+    def align_data_to_axes(cls, gtsfm_data: "GtsfmData", walignedTw: Pose3) -> None:
```
Comment (adisingh50) [inline gtsfm/common/gtsfm_data.py:392]: Mistake on the return type here. Will fix this

**Comment by johnwlambert** [comment]
Thanks for the PR, @adisingh50. Looks like the CI is failing on some tests:
```
=========================== short test summary info ============================
FAILED tests/frontend/verifier/test_verifier_argoverse.py::TestDegensacVerifierArgoverse::testRecoveredPoseError
FAILED tests/utils/test_ellipsoid_utils.py::TestEllipsoidUtils::test_apply_ellipsoid_rotation
FAILED tests/utils/test_ellipsoid_utils.py::TestEllipsoidUtils::test_apply_ellipsoid_rotation_wrong_dims
FAILED tests/utils/test_ellipsoid_utils.py::TestEllipsoidUtils::test_get_rotation_matrix
FAILED tests/utils/test_ellipsoid_utils.py::TestEllipsoidUtils::test_get_rotation_matrix_wrong_dims
======= 5 failed, 272 passed, 4 skipped, 1 warning in 342.62s (0:05:42) ========
```

File: tests/utils/test_ellipsoid_utils.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        expected = np.array([[0, 1, 0], [-1, 0, 0], [0, 0, -1]])
+        npt.assert_almost_equal(computed, expected, 6)
+
+    def test_get_rotation_matrix_wrong_dims(self):
+        """Tests the get_rotation_matrix() function with sample points of 2 dimensions."""
+
+        sample_points = np.array([[1, 1], [2, 2], [-1, -1], [-2, -2], [-1, 1], [1, -1]])
+        self.assertRaises(TypeError, ellipsoid_utils.get_rotation_matrix, sample_points)
+
+    def test_apply_ellipsoid_rotation(self):
```
Comment (johnwlambert) [inline tests/utils/test_ellipsoid_utils.py:74]: Hi Adi, thanks for the PR.
Nit: can you add the return types here?
```python
def test_apply_ellipsoid_rotation(self) -> None
```

File: tests/utils/test_ellipsoid_utils.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        Means of x,y,z is clearly (2,2,2), so centering the point cloud yields:
+            (1,1,1) -> (-1,-1,-1)
+            (2,2,2) -> (0,0,0)
+            (3,3,3) -> (1,1,1)
+        """
+
+        sample_points = np.array([[1, 1, 1], [2, 2, 2], [3, 3, 3]])
+        computed, means = ellipsoid_utils.center_point_cloud(sample_points)
+        expected = np.array([[-1, -1, -1], [0, 0, 0], [1, 1, 1]])
+        npt.assert_almost_equal(computed, expected, 6)
```
Comment (johnwlambert) [inline tests/utils/test_ellipsoid_utils.py:26]: might be nice to use the keyword arg just to increase readability:
```
npt.assert_almost_equal(computed, expected, decimal=6)
```

File: tests/utils/test_ellipsoid_utils.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
+"""Unit tests for functions in ellipsoid utils file.
+
+Authors: Adi Singh
+"""
+import gtsfm.utils.ellipsoid as ellipsoid_utils
```
Comment (johnwlambert) [inline tests/utils/test_ellipsoid_utils.py:5]: let's group these per the [Google style guide](https://google.github.io/styleguide/pyguide.html#313-imports-formatting):

```python
import unittest

import numpy as np
import numpy.testing as npt

import gtsfm.utils.ellipsoid as ellipsoid_utils
```

File: rtf_vis_tool/src/Components/frustum_classes/se3.js [hunk 13]
```diff
@@ -44,7 +44,7 @@ class SE3 {
         var ones = nj.ones([1, 1]);
         var homogeneous_pt = nj.concatenate(point, ones);
 
-        var point_world_coords = homogeneous_pt.dot(this.transform_matrix.T);
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/se3.js:47]: this was right before I believe -- why the change?
```
p_w = wTc * p_c
```
(4,1) = (4,4) * (4,1)
or
p_w = (wTc * p_c)^T
p_w^T = p_c^T (wTc)^T
(1,4) = (1,4) * (4,4)
See Boyd's Crimes against Matrices: http://ee263.stanford.edu/notes/matrix_crimes.pdf

So we can do
(N,4) = (N,4) * (4,4)

However, we actually dont need the homogenous thing at all, we can just do 
```python
return point_cloud @ self.rotation.T + self.translation
```

Do we have a unit test on this?

File: rtf_vis_tool/src/Components/AllFrustums.js [hunk 14]
```diff
@@ -98,8 +99,8 @@ function AllFrustums() {
             const rotation_matrix = q.toMatrix(true);
             const rotation = nj.array(rotation_matrix);
             const translation = nj.array([tx,ty,tz]);
-            const wTc = new SE3(rotation, translation);
-            var verts_worldframe = frustumObj.get_mesh_vertices_worldframe(wTc);
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/AllFrustums.js:102]: let's keep it the way I had in my PR: https://github.com/borglab/gtsfm/pull/297/files

All we need to do is add the inverse operation to the SE(3) class, and call it here.

File: gtsfm/common/gtsfm_data.py [hunk 15]
```diff
@@ -387,3 +387,53 @@ def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "Gtsfm
...
+        """
+        if walignedTw is None:
+            raise ValueError("walignedTw cannot be None, should be a valid alignment transformation.")
+
+        num_tracks = gtsfm_data.number_tracks()
+        aligned_data = cls(gtsfm_data.number_images())
+
+        # Populate aligned_gtsfm_data with duplicate camera data.
+        for camera_index in gtsfm_data.get_valid_camera_indices():
+            aligned_data.add_camera(camera_index, gtsfm_data.get_camera(camera_index))
```
Comment (johnwlambert) [inline gtsfm/common/gtsfm_data.py:410]: it looks like we are adding the cameras twice, in line 410 and in line 437. let's just do it once.

File: gtsfm/common/gtsfm_data.py [hunk 16]
```diff
@@ -387,3 +387,53 @@ def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "Gtsfm
...
+
+        # Populate aligned_data with tracks.
+        for i in range(num_tracks):
+            original_track_3d = gtsfm_data.get_track(i)
+            point = original_track_3d.point3()
+            aligned_point = walignedTw.matrix() @ np.append(point, [1], 0)
```
Comment (johnwlambert) [inline gtsfm/common/gtsfm_data.py:416]: we can just use `aligned_point = walignedTw.transformFrom(point)` like we do in line 380.

File: gtsfm/common/gtsfm_data.py [hunk 17]
```diff
@@ -387,3 +387,53 @@ def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "Gtsfm
             aligned_data.add_track(track_aligned)
 
         return aligned_data
+
+    @classmethod
+    def align_data_to_axes(cls, gtsfm_data: "GtsfmData", walignedTw: Pose3) -> None:
+        """Align the GtsfmData's tracks and poses to the x,y,z axes.
+
+        Args:
```
Comment (johnwlambert) [inline gtsfm/common/gtsfm_data.py:395]: if we rewrite the function above as:
```python
    def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "GtsfmData":
        """Align estimated, sparse multiview result (self) to a set of reference poses.
        Args:
            wTi_list_ref: list of reference/target camera poses, ordered by camera index.
        Returns:
            aligned_data: sparse multiview result that is aligned to the poses above.
        """
        # these are the estimated poses (source, to be aligned)
        wTi_list = self.get_camera_poses()
        # align the poses which are valid (i.e. are not None)
        # some camera indices may have been lost after pruning to largest connected component, leading to None values
        # rSe aligns the estimate `e` frame to the reference `r` frame
        wTi_list_aligned, rSe = geometry_comparisons.align_poses_sim3_ignore_missing(wTi_list_ref, wTi_list)
        return self.apply_Sim3(aSb=rSe)

    def apply_Sim3(self, aSb: Sim3) -> "GtsfmData" 
       aligned_data = GtsfmData(number_images=self.number_images())
        # update the camera pose to the aligned poses, but use the previous calibration
        for i, wTi in enumerate(wTi_list_aligned):
            if wTi is None:
                continue
            calibration = self.get_camera(i).calibration()
            aligned_data.add_camera(i, PinholeCameraCal3Bundler(wTi, calibration))
        # align estimated tracks to the ground truth
        for j in range(self.number_tracks()):
            # align each 3d point
            track_est = self.get_track(index=j)
            # place into the GT reference frame
            pt_ref = rSe.transformFrom(track_est.point3())
            track_aligned = SfmTrack(pt_ref)
            # copy over the 2d measurements directly into the new track
            for k in range(track_est.number_measurements()):
                i, uv = track_est.measurement(k)
                track_aligned.add_measurement(i, uv)
            aligned_data.add_track(track_aligned)

        return aligned_data
```
then this should be a 1 liner as follows:
```python
walignedTw = Similarity3(R=walignedTw.rotation(), t=walignedTw.translation(), s=1.0)
ba_output.apply_Sim3(walignedTw)
```
It looks like all of the code is pretty much the same.
```

File: rtf_vis_tool/src/Components/frustum_classes/se3.js (hunk 13 above)
Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:47]: Thanks for breaking this down and sending the link. Not sure why I changed it, but I've reverted the change. I do not have unit tests for `se3.js`, so I will add those in.

File: gtsfm/common/gtsfm_data.py (hunk 17 above)
Comment (adisingh50) [reply gtsfm/common/gtsfm_data.py:395]: Thanks for writing out all this code for me :). I will update `align_via_Sim3_to_poses` in `gtsfm_data.py` to what you have and then add your new method called `apply_Sim3` to `gtsfm_data.py`

**Comment by adisingh50** [comment]
Sorry for the extra commits. Had some weird git issues that led me to push more commits than I should have.

File: gtsfm/utils/ellipsoid.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+"""
+
+import numpy as np
+from typing import Tuple
+from gtsam import Pose3, Rot3
+
+from gtsfm.common.gtsfm_data import GtsfmData
+
+
+def transform_point_cloud_wrapper(gtsfm_data: GtsfmData) -> Pose3:
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:14]: thanks for all the updates Adi, looks quite good. Almost ready for merge I think.

Let's try to think of a more descriptive name for this function, since `wrapper` is a bit vague.

File: rtf_vis_tool/src/Components/frustum_classes/se3.js [hunk 19]
```diff
@@ -51,6 +51,22 @@ class SE3 {
...
     }
+
+    inverse() {
+        /* Computes the inverse of the current pose matrix. (Since numjs doesn't have a function for it)
+
+        Returns:
+            A new SE3 object storing the inverse of the current pose matrix.
+        */
+        const Rt = this.rotation.T;
+        const t_negative = nj.array([-1 * this.translation.get(0),
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/se3.js:62]: is it not possible to just multiply the whole array by -1, instead of doing it on a per-element basis?

File: rtf_vis_tool/src/Components/frustum_classes/se3.js [hunk 20]
```diff
@@ -51,6 +51,22 @@ class SE3 {
...
+            A new SE3 object storing the inverse of the current pose matrix.
+        */
+        const Rt = this.rotation.T;
+        const t_negative = nj.array([-1 * this.translation.get(0), 
+                              -1 * this.translation.get(1), 
+                              -1 * this.translation.get(2)]);
+
+        const t_inverse = Rt.dot(t_negative);
+        const inverse = new SE3(Rt, t_inverse);
+        return inverse;
```
Comment (johnwlambert) [inline rtf_vis_tool/src/Components/frustum_classes/se3.js:68]: I think this would be cleaner just as
```js
return new SE3(Rt, Rt.dot(t_negative))
```

Comment (johnwlambert) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:68]: or if we want the const, which we probably do, i would prefer:
```js
const inverse = new SE3(Rt, Rt.dot(t_negative));
return inverse;
```

File: rtf_vis_tool/src/tests/SE3.test.js [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+// Local Imports.
+var SE3 = require('../Components/frustum_classes/se3');
+
+describe('SE3 Class Unit Tests', () => {
+
+    /* Creates a simple Pose matrix that flips x and z coordinates and translates by [2,3,4]. 
+       Applies to the sample point [1,2,3] and verifies the SE3 transformation result is accurate.
+    */
+    it('Test transform_from.', () => {
+        const rotation = nj.array([[0, 0, 1],
```
Comment (johnwlambert) [inline rtf_vis_tool/src/tests/SE3.test.js:18]: maybe a comment could explain that this rotation matrix corresponds to Rx, Ry, Rz of [180, 0, 180] degrees?

File: rtf_vis_tool/src/tests/SE3.test.js [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+                                   [0, 1, 0],
+                                   [1, 0, 0]]);
+        const translation = nj.array([2,3,4]);
+        const testSE3 = SE3(rotation, translation);
+        const point = nj.array([1, 2, 3]);
+
+        const computedPoint = testSE3.transform_from(point);
+        const expectedPoint = nj.array([[5, 5, 5]]);
```
Comment (johnwlambert) [inline rtf_vis_tool/src/tests/SE3.test.js:26]: looks good, thanks for adding this

File: tests/utils/test_ellipsoid_utils.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    def test_remove_outlier_points_wrong_dims(self) -> None:
+        """Tests the remove_outlier_points() function with 5 sample points of 2 dimensions."""
+
+        sample_points = np.array([[6, 13], [5, 9], [6, 10], [9, 13], [5, 12]])
+        self.assertRaises(TypeError, ellipsoid_utils.remove_outlier_points, sample_points)
+
+    def test_get_alignment_rotation_matrix_from_svd(self) -> None:
+        """Tests the get_alignment_rotation_matrix_from_svd() function with 6 sample points."""
+
+        sample_points = np.array([[1, 1, 0], [-1, 1, 0], [-2, 2, 0], [-1, -1, 0], [1, -1, 0], [2, -2, 0]])
```
Comment (johnwlambert) [inline tests/utils/test_ellipsoid_utils.py:61]: I think this would be a littler easier to parse visually as:
```python
fmt: off
sample_points = np.array(
    [
        [1, 1, 0],
        [-1, 1, 0],
        [-2, 2, 0],
        [-1, -1, 0],
        [1, -1, 0],
        [2, -2, 0]
    ]
)

fmt: on
```

File: gtsfm/scene_optimizer.py [hunk 24]
```diff
@@ -216,6 +219,11 @@ def _filter_dict_keys(dict: Dict[Any, Any], ref_dict: Dict[Any,Any]) -> Dict[Any
         # Save metrics to JSON and generate HTML report.
         auxiliary_graph_list.extend(save_metrics_reports(metrics_graph_list))
 
+        # Modify BA input and BA output to have point clouds and frustums aligned with x,y,z axes.
+        aligned_pose_graphs = dask.delayed(align_estimated_gtsfm_data)(ba_input_graph, ba_output_graph)
+        ba_input_graph = aligned_pose_graphs[0]
+        ba_output_graph = aligned_pose_graphs[1]
```
Comment (johnwlambert) [inline gtsfm/scene_optimizer.py:225]: can you post an example of the output matplotlib plots with this code change? we need to apply the transformation to align to orthogonal axes to the ground truth here, otherwise the plotting code will look incorrect.

Comment (johnwlambert) [reply gtsfm/scene_optimizer.py:225]: this code will be cleaner with 
```python
ba_input_graph, ba_output_graph = ask.delayed(align_estimated_gtsfm_data, nout=2)(ba_input_graph, ba_output_graph)
```

File: gtsfm/utils/ellipsoid.py [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+    mags = np.linalg.norm(point_cloud, axis=1)
+    cutoff_mag = np.percentile(mags, 95)
+    points_filtered = point_cloud[mags < cutoff_mag]
+    return points_filtered
+
+
+def get_alignment_rotation_matrix_from_svd(point_cloud: np.ndarray) -> np.ndarray:
+    """Applies SVD on a point cloud. The resulting V contains the rotation matrix required to align the 3 principal
+    axes of the point cloud's ellipsoid with the x, y, z coordinate axes.
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:89]: i think it would be clearer to say "to fit an ellipsoid to the point cloud" rather than to say the "point cloud's ellipsoid", since we haven't defined that yet.

File: gtsfm/utils/ellipsoid.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+        The filtered point cloud of shape M x 3 (M = 0.95*N).
+
+    Raises:
+        TypeError: if centered point cloud is not of shape (N,3).
+    """
+    if point_cloud.shape[1] != 3:
+        raise TypeError("Point Cloud should be 3 dimensional")
+
+    mags = np.linalg.norm(point_cloud, axis=1)
+    cutoff_mag = np.percentile(mags, 95)
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:82]: i would make 95 a global constant in this module, at the top, like
```
OUTLIER_DISTANCE_PERCENTILE = 95
```

File: gtsfm/utils/ellipsoid.py [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+        gtsfm_data: scene data to write to transform.
+
+    Returns:
+        The final transformation required to align point cloud and frustums.
+    """
+    # Iterate through each track to gather a list of 3D points forming the point cloud.
+    point_cloud_list = []
+    num_pts = gtsfm_data.number_tracks()
+
+    for j in range(num_pts):
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:28]: maybe it's simpler to just write:
```python
point_cloud = [gtsfm_data.get_track(j).point3() for for j in range(num_pts)]
point_cloud = np.array(point_cloud)
```

File: gtsfm/utils/ellipsoid.py [hunk 28]
```diff
@@ -0,0 +1,108 @@
+"""Algorithms to center and align 3D points and camera frustums to the x, y, and z axes using SVD. Used in React
+Three Fiber Visulization Tool.
+
+Authors: Adi Singh
+"""
+
+import numpy as np
+from typing import Tuple
```
Comment (johnwlambert) [inline gtsfm/utils/ellipsoid.py:8]: imports should be ordered a bit differently:
```python
from typing import Tuple

import numpy as np
from gtsam import Pose3, Rot3

from gtsfm.common.gtsfm_data import GtsfmData
```

File: gtsfm/scene_optimizer.py [hunk 29]
```diff
@@ -231,6 +239,25 @@ def _filter_dict_keys(dict: Dict[Any, Any], ref_dict: Dict[Any,Any]) -> Dict[Any
         return output_graph[0]
 
 
+def align_estimated_gtsfm_data(ba_input: GtsfmData, ba_output: GtsfmData) -> Tuple[Delayed]:
```
Comment (johnwlambert) [inline gtsfm/scene_optimizer.py:242]: so the return type should be 
```python
def align_estimated_gtsfm_data(ba_input: GtsfmData, ba_output: GtsfmData) -> Tuple[GtsfmData, GtsfmData]
```
this is because the function call is wrapped with `dask.delayed()`. We only mark the return type as dask.Dalayed when the function call is not wrapped.

File: gtsfm/common/gtsfm_data.py [hunk 30]
```diff
@@ -358,32 +358,39 @@ def align_via_Sim3_to_poses(self, wTi_list_ref: List[Optional[Pose3]]) -> "Gtsfm
...
         wTi_list = self.get_camera_poses()
-
         # align the poses which are valid (i.e. are not None)
         # some camera indices may have been lost after pruning to largest connected component, leading to None values
         # rSe aligns the estimate `e` frame to the reference `r` frame
-        wTi_list_aligned, rSe = geometry_comparisons.align_poses_sim3_ignore_missing(wTi_list_ref, wTi_list)
+        _, rSe = geometry_comparisons.align_poses_sim3_ignore_missing(wTi_list_ref, wTi_list)
+        return self.apply_Sim3(aSb=rSe)
+
+    def apply_Sim3(self, aSb: Similarity3) -> "GtsfmData":
```
Comment (johnwlambert) [inline gtsfm/common/gtsfm_data.py:367]: looks good.

File: gtsfm/utils/ellipsoid.py (hunk 18 above)
Comment (adisingh50) [reply gtsfm/utils/ellipsoid.py:14]: Ok, I'm changing it to `get_point_cloud_alignment_transform` since the function is not even transforming the point cloud anymore. It's just calculating and returning `walignedTw`

Comment (johnwlambert) [reply gtsfm/utils/ellipsoid.py:14]: how about `get_ortho_axis_alignment_transform()` ? (since we're aligning both camera poses and points)

File: rtf_vis_tool/src/Components/frustum_classes/se3.js (hunk 20 above)
Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:68]: yup that looks cleaner

File: rtf_vis_tool/src/Components/frustum_classes/se3.js (hunk 19 above)
Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:62]: Unfortunately, `numjs` does not support broadcasting, so it won't allow me to multiply an array by a scalar.

Comment (adisingh50) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:62]: https://github.com/nicolaspanel/numjs/issues/1

Comment (johnwlambert) [reply rtf_vis_tool/src/Components/frustum_classes/se3.js:62]: cool, nevermind then

File: gtsfm/utils/ellipsoid.py (hunk 18 above)
Comment (adisingh50) [reply gtsfm/utils/ellipsoid.py:14]: Sounds good to me

File: gtsfm/scene_optimizer.py (hunk 24 above)
Comment (adisingh50) [reply gtsfm/scene_optimizer.py:225]: Slacked you the matplotlib plots of all the camera poses. Just added this line under `align_estimated_gtsfm_data()` method:
`gt_pose_graph = [walignedTw.transformFrom(wTi) for wTi in gt_pose_graph]`

Comment (johnwlambert) [reply gtsfm/scene_optimizer.py:225]: thanks!

**Review by johnwlambert — APPROVED** [review]
LGTM
